File: plugins/sample-java/src/main/java/com/example/plugin/AdBlockerPlugin.java
package com.example.plugin;

import android.util.Log;
import com.fluxfabric.model.Packet;
import com.fluxfabric.plugin.BasePlugin;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

public class AdBlockerPlugin extends BasePlugin {

    private static final String TAG = "AdBlocker";
    private int blockedCount = 0;

    private static final Set<String> BLOCKED_DOMAINS = new HashSet<>(Arrays.asList(
        "doubleclick.net",
        "googlesyndication.com",
        "googleadservices.com",
        "adservice.google.com",
        "adsrvr.org",
        "adnxs.com",
        "rubiconproject.com"
    ));

    public AdBlockerPlugin() {
        super(
            "ad-blocker",
            "Ad Blocker",
            "Blocks known ad and tracking domains",
            "1.0.0"
        );
    }

    @Override
    public void onStart() {
        blockedCount = 0;
        Log.d(TAG, "AdBlocker started with " + BLOCKED_DOMAINS.size() + " blocked domains");
    }

    @Override
    public void onStop() {
        Log.d(TAG, "AdBlocker stopped — blocked " + blockedCount + " requests");
    }

    @Override
    public Packet onPacketSent(Packet packet) {
        // Check DNS requests (UDP port 53)
        if (packet.getProtocol() == Packet.Protocol.UDP && packet.getDstPort() == 53) {
            String dnsQuery = extractDnsQuery(packet.getData());
            if (dnsQuery != null && isBlocked(dnsQuery)) {
                blockedCount++;
                Log.d(TAG, "Blocked DNS query to: " + dnsQuery);
                return null; // Drop the packet
            }
        }
        return packet;
    }

    private String extractDnsQuery(byte[] data) {
        // Simple DNS query extraction (skips IP + UDP headers)
        try {
            java.nio.ByteBuffer bb = java.nio.ByteBuffer.wrap(data);
            int versionAndIhl = bb.get() & 0xFF;
            int ihl = (versionAndIhl & 0x0F) * 4;
            int udpOffset = ihl + 8; // IP header + UDP header

            if (data.length <= udpOffset + 12) return null;

            // Skip DNS header (12 bytes)
            int queryStart = udpOffset + 12;
            StringBuilder domain = new StringBuilder();

            int pos = queryStart;
            while (pos < data.length) {
                int len = data[pos] & 0xFF;
                if (len == 0) break;
                pos++;
                if (pos + len > data.length) break;
                if (domain.length() > 0) domain.append(".");
                for (int i = 0; i < len; i++) {
                    domain.append((char) data[pos + i]);
                }
                pos += len;
            }

            return domain.length() > 0 ? domain.toString().toLowerCase() : null;
        } catch (Exception e) {
            return null;
        }
    }

    private boolean isBlocked(String domain) {
        for (String blocked : BLOCKED_DOMAINS) {
            if (domain.endsWith(blocked)) {
                return true;
            }
        }
        return false;
    }
}

File: src/main/java/com/fluxfabric/plugin/builtin/AdBlockerPlugin.kt
package com.fluxfabric.plugin.builtin

import android.util.Log
import com.fluxfabric.model.Packet
import com.fluxfabric.plugin.BasePlugin
import com.fluxfabric.plugin.PluginSetting

class AdBlockerPlugin : BasePlugin(
    id = "ad-blocker",
    name = "Ad Blocker",
    description = "Blocks known ad/tracking domains via DNS filtering",
    version = "1.0.0"
) {
    private var blockedCount = 0

    override val settings: List<PluginSetting<*>> = listOf(
        PluginSetting.BooleanSetting(
            key = "block_ads",
            label = "Block Ad Domains",
            description = "Block known advertising and tracking domains",
            defaultValue = true
        ),
        PluginSetting.StringSetting(
            key = "custom_domains",
            label = "Custom Domains",
            description = "Additional domains to block (comma-separated)",
            defaultValue = ""
        ),
        PluginSetting.IntSetting(
            key = "max_blocked",
            label = "Max Blocked",
            description = "Maximum number of blocked requests before pausing (0 = unlimited)",
            defaultValue = 0,
            min = 0,
            max = 10000
        )
    )

    private val builtinDomains = setOf(
        "doubleclick.net", "googlesyndication.com", "googleadservices.com",
        "adservice.google.com", "adsrvr.org", "adnxs.com", "rubiconproject.com",
        "adform.net", "criteo.com", "casalemedia.com", "moatads.com",
        "outbrain.com", "taboola.com", "pubmatic.com", "openx.net",
        "appnexus.com", "adzerk.net", "scorecardresearch.com", "quantserve.com",
        "exelator.com", "bluekai.com", "demdex.net"
    )

    private val activeDomains: Set<String>
        get() {
            val domains = mutableSetOf<String>()
            if (getSettingValue("block_ads") == true) {
                domains.addAll(builtinDomains)
            }
            val custom = (getSettingValue("custom_domains") as? String)?.takeIf { it.isNotBlank() }
            if (custom != null) {
                domains.addAll(custom.split(",").map { it.trim() }.filter { it.isNotEmpty() })
            }
            return domains
        }

    override fun onStart() {
        blockedCount = 0
        Log.d(TAG, "AdBlocker started (${activeDomains.size} domains)")
    }

    override fun onStop() {
        Log.d(TAG, "AdBlocker stopped — blocked $blockedCount requests")
    }

    override fun onPacketSent(packet: Packet): Packet? {
        val max = (getSettingValue("max_blocked") as? Int) ?: 0
        if (max > 0 && blockedCount >= max) return packet
        if (packet.protocol == Packet.Protocol.UDP && packet.dstPort == 53) {
            val domain = extractDnsQuery(packet.data)
            if (domain != null && isBlocked(domain)) {
                blockedCount++
                Log.i(TAG, "Blocked DNS: $domain")
                return null
            }
        }
        return packet
    }

    private fun extractDnsQuery(data: ByteArray): String? {
        return try {
            val bb = java.nio.ByteBuffer.wrap(data)
            val versionAndIhl = bb.get().toInt() and 0xFF
            val ihl = (versionAndIhl and 0x0F) * 4
            val udpOffset = ihl + 8
            if (data.size <= udpOffset + 12) return null
            val queryStart = udpOffset + 12
            val domain = StringBuilder()
            var pos = queryStart
            while (pos < data.size) {
                val len = data[pos].toInt() and 0xFF
                if (len == 0) break
                pos++
                if (pos + len > data.size) break
                if (domain.isNotEmpty()) domain.append(".")
                for (i in 0 until len) {
                    domain.append((data[pos + i].toInt() and 0xFF).toChar())
                }
                pos += len
            }
            domain.toString().lowercase().ifEmpty { null }
        } catch (_: Exception) { null }
    }

    private fun isBlocked(domain: String): Boolean {
        return activeDomains.any { domain.endsWith(it) }
    }

    companion object {
        private const val TAG = "AdBlocker"
    }
}
